shim.go

```diff
@@ -160,17 +166,6 @@ func (s *shim) resizeTty(fromTty *os.File) error {
 	return err
 }
 
-func (s *shim) monitorTtySize(tty *os.File) {
-	s.resizeTty(tty)
-	sigchan := make(chan os.Signal, 1)
-	signal.Notify(sigchan, syscall.SIGWINCH)
```

Comment:
I thought the original code handles SIGWINCH properly. Then the purpose of this PR is to move SIGWINCH handling to the unified place, is that correct? If so, please create a different issue for it and we can close #74 because it is not valid for kata-shim.

Reply:
Hi @bergwolf - Sorry for the confusion! Yes, this PR is fixing a different problem to #74. This PR allows a `SIGWINCH` sent direct to the shim to be passed to the container workload, which does *not* work currently.

I've also raised https://github.com/kata-containers/tests/issues/550 since #74 had been left open (incorrectly - now closed :)

Reply:
@jodh-intel I guess another question is that, do we need `SIGWINCH` to be passed to the guest? I think the resizeTTY grpc is designed to replace `SIGWINCH`. And since you still call `s.resizeTty(tty)` in case of `SIGWINCH`, what benefit does it bring in to forward the signal to the guest?

It seems by doing so, the guest process would have to resize tty twice (one for s.resizeTty(tty) and another for SIGWINCH). And I'm not sure what tty size the process can pick when getting SIGWINCH because the tty size on the host is not passed along with it.

Reply:
Hi @bergwolf - I understand what you're saying but fundamentally, the shim represents the real workload, so all signals sent to it should be sent to the real workload (where possible).

Any thoughts @sboeuf?

Reply:
Ok so if I understand this PR correctly, this is a simple refactoring, which looks very sane to me, of having SIGWINCH handled from the main loop handling all the signals. There's no harm doing so as I think it does not modify the current behavior.

One comment though, `s.resizeTty(tty)` was called a first time when the shim was started, but this PR only relies on signals now. We might want to keep this call to make sure the container inside the VM has the right winsize even if SIGWINCH was never sent to the shim.

Reply:
@sboeuf Take a look at here https://github.com/kata-containers/shim/pull/100/files#diff-82603f885ec5ee584c3502e9fc5e43f9R129. It is not simply refactoring. It also actually sends `SIGWINCH` to the agent. It'll be great if this PR just refactors to put signal handling in a single place and do not change the current behavior (`SIGWINCH` is translated in to `s.resizeTty` and not sent to the agent).

Reply:
Oh yes I can see that now. The new behavior being, every time the shim receives a SIGWINCH, it will first resize the container winsize by using the agent protocol, and in a second time, it will signal the container process by forwarding the SIGWINCH signal.
I have to say I might not be clear about the expected semantics here, but I think @jodh-intel might be right with this PR actually.
If you think about a standard runc container, when docker decide to send SIGWINCH, well the expected behavior is that SIGWINCH will notify the container process that the window changed. And if you think about it, the window did change because docker (or whoever the caller is) modified the terminal size, and then sent the signal to the container.
Now, if we're thinking about mimicking this behavior in kata, when the shim receives the signal, it has to look for the new window size, update it in the VM by using the agent protocol, which will apply the settings to the container terminal inside the VM, and finally will have to notify the "real" container process that the window has changed. This is the part where we actually need to forward the SIGWINCH signal.
I'm realizing now that @jodh-intel's approach makes a lot of sense !

Reply:
@sboeuf @jodh-intel I still don't think we need to pass the `SIGWINCH` signal to the agent. The `SIGWINCH` signal is sent to processes by the guest kernel when the window size is changed. There is no need to send another `SIGWINCH` to the process by shim. See man tty_ioctl(4):
```
       TIOCSWINSZ     const struct winsize *argp
              Set window size.

       The struct used by these ioctls is defined as

           struct winsize {
               unsigned short ws_row;
               unsigned short ws_col;
               unsigned short ws_xpixel;   /* unused */
               unsigned short ws_ypixel;   /* unused */
           };

       When the window size changes, a SIGWINCH signal is sent to the foreground process group.
```

Reply:
@jodh-intel could you please give more feedback on this ?